core/survey/_survey/surveyNodeDefsLayout.js

```diff
@@ -0,0 +1,36 @@
+import * as R from 'ramda'
+
+import { uuidv4 } from '@core/uuid'
+import * as NodeDef from '@core/survey/nodeDef'
+import * as NodeDefLayout from '@core/survey/nodeDefLayout'
+import * as SurveyNodeDefs from './surveyNodeDefs'
+
+export const updateNodeDefLayoutProp = ({ surveyCycleKey, nodeDef, key, value }) => (survey) => {
+  let nodeDefLayout = R.pipe(
+    NodeDefLayout.getLayout,
+    NodeDefLayout.assocLayoutProp(surveyCycleKey, key, value)
+  )(nodeDef)
+
+  // If setting layout render mode (table | form), set the the proper layout
+  if (NodeDef.isEntity(nodeDef) && key === NodeDefLayout.keys.renderType) {
+    if (value === NodeDefLayout.renderType.table) {
+      // Render mode table
+      // Assoc layout children
+      const nodeDefChildren = SurveyNodeDefs.getNodeDefChildren(nodeDef)(survey)
+      nodeDefLayout = NodeDefLayout.assocLayoutChildren(
+        surveyCycleKey,
+        R.map(NodeDef.getUuid, nodeDefChildren)
+      )(nodeDefLayout)
+    } else {
+      // Render mode form
+      // Dissoc layoutChildren (applicable only if render mode is table)
+      nodeDefLayout = NodeDefLayout.dissocLayoutChildren(surveyCycleKey)(nodeDefLayout)
+      // Entity rendered as form can only exists in its own page
+      if (NodeDefLayout.isDisplayInParentPage(surveyCycleKey)(nodeDef)) {
+        nodeDefLayout = NodeDefLayout.assocPageUuid(surveyCycleKey, uuidv4())(nodeDefLayout)
+      }
+    }
+  }
+  const nodeDefUpdated = NodeDefLayout.assocLayout(nodeDefLayout)(nodeDef)
+  return SurveyNodeDefs.assocNodeDef(nodeDefUpdated)(survey)
+}
```

Comment:
I find this method difficulty to understand and the comments can be a good "code smell" 

Some articles from Martin Fowler and Uncle Bob talking about that:
[https://blog.usejournal.com/stop-writing-code-comments-28fef5272752](https://blog.usejournal.com/stop-writing-code-comments-28fef5272752)
[https://martinfowler.com/bliki/CodeAsDocumentation.html]
(https://martinfowler.com/bliki/CodeAsDocumentation.html)
[https://blog.cleancoder.com/uncle-bob/2017/02/23/NecessaryComments.html](https://blog.cleancoder.com/uncle-bob/2017/02/23/NecessaryComments.html)

```
const _nodeDefLayoutUpdatedByRenderType = ({ surveyCycleKey, nodeDef, value, nodeDefLayout }) => {    
    if(value === NodeDefLayout.renderType.table){
      // Assoc layout children
      const nodeDefChildren = SurveyNodeDefs.getNodeDefChildren(nodeDef)(survey)
      return NodeDefLayout.assocLayoutChildren(
        surveyCycleKey,
        R.map(NodeDef.getUuid, nodeDefChildren)
      )(nodeDefLayout)
    }

    if(value === NodeDefLayout.renderType.form){
      let nodeDefLayoutUpdated = NodeDefLayout.dissocLayoutChildren(surveyCycleKey)(nodeDefLayout)
      // Entity rendered as form can only exists in its own page
      if (NodeDefLayout.isDisplayInParentPage(surveyCycleKey)(nodeDef)) {
        nodeDefLayoutUpdated = NodeDefLayout.assocPageUuid(surveyCycleKey, uuidv4())(nodeDefLayoutUpdated)
      }
      return nodeDefLayoutUpdated
    }

    return nodeDefLayout
}

const _isNodeDefEntityAndKeyRenderMode = ({ nodeDef, key }) => NodeDef.isEntity(nodeDef) && key === NodeDefLayout.keys.renderType

export const updateNodeDefLayoutProp = ({ surveyCycleKey, nodeDef, key, value }) => (survey) => {
  let nodeDefLayout = R.pipe(
    NodeDefLayout.getLayout,
    NodeDefLayout.assocLayoutProp(surveyCycleKey, key, value)
  )(nodeDef)

  if (_isNodeDefEntityAndKeyRenderMode({nodeDef, key}))
    nodeDefLayout = _nodeDefLayoutUpdatedByRenderType({ surveyCycleKey, nodeDef, key, value, nodeDefLayout })
  
  const nodeDefUpdated = NodeDefLayout.assocLayout(nodeDefLayout)(nodeDef)
  return SurveyNodeDefs.assocNodeDef(nodeDefUpdated)(survey)
}
```

my proposal can be improved, it is only an idea about how to improve the implementation, in fact, the `_nodeDefLayoutUpdatedByRenderType` has room for improvement

Reply:
yes thanks, it's already a good starting point to improve it, I'll change it into this.

Reply:
Thanks!